### PR TITLE
online-from and online-to cannot be empty

### PR DESCRIPTION
**Pull request by maxakropolis**
In the catalog.xsd online-from and onlne-to cannot be empty

```
<xsd:element name="online-from" type="sharedType.SiteSpecificDateTime" minOccurs="0" maxOccurs="unbounded" nillable="true" />
<xsd:element name="online-to" type="sharedType.SiteSpecificDateTime" minOccurs="0" maxOccurs="unbounded" nillable="true" />
```

However they can be set to nil `<online-from xsi:nil="true"/>`

### CODE REVIEW

File: src/Writer/Xml/NilEmptyWriter.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+class NilEmptyWriter
+{
+    private $writer;
+
+    public function __construct(XmlWriter $writer)
+    {
+        $this->writer = $writer;
+    }
+
+    public function writeElement($name, $content = null): bool
```
Comment (dannorcott) [inline src/Writer/Xml/NilEmptyWriter.php:13]: Is name deliberately not type hinted?

Comment (maxakropolis) [reply src/Writer/Xml/NilEmptyWriter.php:13]: Yeah - I copied `NilEmptyWriter` from `NotEmptyWriter`

Comment (dannorcott) [reply src/Writer/Xml/NilEmptyWriter.php:13]: Ah yeah, I dragged it down and see they're just aping the methods they're calling

File: src/Writer/Xml/NilEmptyWriter.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        $this->writer->startElement($name);
+
+        foreach ($attributes as $attrName => $attrContent) {
+            $this->writeAttribute($attrName, $attrContent);
+        }
+
+        if (XmlFormatter::isEmptyValue($content)) {
+            $this->writeAttribute('xsi:nil', 'true');
+            $this->writeAttribute('xmlns:xsi', 'http://www.w3.org/2001/XMLSchema-instance');
```
Comment (dannorcott) [inline src/Writer/Xml/NilEmptyWriter.php:37]: Suspect out of scope but http://www.w3.org/2001/XMLSchema-instance is repeated a fair bit around the place and might benefit from centralisation

File: tests/Writer/CategoriesTest.php [hunk 3]
```diff
@@ -19,22 +19,46 @@ public function test_categories_xml(): void
...
+                'dates' => [
+                    null,
+                    new DateTimeImmutable('2018-02-02 02:02:02'),
+                ],
+            ],
+            'Donuts' => [
+                'index' => 2,
+                'dates' => [
+                    new DateTimeImmutable('2018-01-01 01:01:01'),
+                    null,
```
Comment (dannorcott) [inline tests/Writer/CategoriesTest.php:41]: as it's just nilIfEmpty rather than nilIfNull, is it worth an empty string test?

**Review by dannorcott — APPROVED** [review]
Seems good to me, apart from my inane drivel

File: src/Writer/Xml/NilEmptyWriter.php (hunk 2 above)
Comment (maxakropolis) [reply src/Writer/Xml/NilEmptyWriter.php:37]: I think it's only used in one place programmatically - the rest of the uses are tests and I'm more comfortable with them being strings.

File: tests/Writer/CategoriesTest.php (hunk 3 above)
Comment (maxakropolis) [reply tests/Writer/CategoriesTest.php:41]: Added some extra tests

File: src/Writer/Xml/NilEmptyWriter.php (hunk 2 above)
Comment (dannorcott) [reply src/Writer/Xml/NilEmptyWriter.php:37]: Oh aye in that case I shall shut my fool mouth

File: tests/Writer/CategoriesTest.php [hunk 4]
```diff
@@ -19,22 +19,46 @@ public function test_categories_xml(): void
...
         $xml->startCatalog('TestCatalog');
 
-        foreach (['Socks', 'Death Stars', 'Donuts'] as $index => $example) {
-            $element = new Category('CAT' . $index);
-            $element->setDisplayName($example);
+        $categories = [
+            'Socks' => [
+                'index' => 0,
+                'dates' => [
+                    new DateTimeImmutable('2018-01-01 01:01:01'),
```
Comment (ziadoz) [inline tests/Writer/CategoriesTest.php:26]: Be nice to give these arrays keys (`from` and `to`).

**Review by ziadoz — APPROVED** [review]
One minor comment about array keys to make the test slightly easier to read, but otherwise this all makes sense.

File: tests/Writer/CategoriesTest.php (hunk 4 above)
Comment (maxakropolis) [reply tests/Writer/CategoriesTest.php]: Sorted

**Pull request merged**